ArchitectureFlow/ArchitectureFlow/Login/UseCase/LoginInteractor.m

```diff
@@ -0,0 +1,102 @@
+#import "LoginInteractor.h"
+#import "Encryption.h"
+#import "EncryptionGateway.h"
+#import "Promise.h"
+#import "LoginRequestModel.h"
+#import "LoginOutput.h"
+#import "UserSession.h"
+#import "UserSessionGateway.h"
+#import "UserSessionLoginRequestModel.m"
+#import "LoginRequestToUserSessionMapper.h"
+#import "TrackingGateway.h"
+
+@interface LoginInteractor ()
+
+@property (nonatomic) id <EncryptionGateway>encryptionGateway;
+@property (nonatomic) id <UserSessionGateway>userSessionGateway;
+@property (nonatomic) id <TrackingGateway>trackingGateway;
+@property (nonatomic) LoginRequestToUserSessionMapper *requestToUserSessionMapper;
+
+@end
+
+@implementation LoginInteractor
+
+- (instancetype)initWithEncryptionGateway:(id <EncryptionGateway>)encryptionGateway userSessionGateway:(id <UserSessionGateway>)userSessionGateway loginRequestToUserSessionMapper:(LoginRequestToUserSessionMapper *)loginRequestToUserSessionMapper trackingGateway:(id <TrackingGateway>)trackingGateway {
+    self = [super init];
+    if (self) {
+        _encryptionGateway = encryptionGateway;
+        _userSessionGateway = userSessionGateway;
+        _requestToUserSessionMapper = loginRequestToUserSessionMapper;
```

Comment:
I'm thinking how could we avoid suddenly being flooded with 1000s of classes. This mapper seems to do a pretty trivial thing in this example. Can this be optional when trivial and when its only used in 1 occurence? Or maybe can the mapping be part of the model object? Is the mapper something that we would mock?

Reply:
1) Avoid being flooded with 1000s of classes: I think we need to take a step back. I've implemented clean architecture many times in the past. Each time there is this concern. This is because the clean architecture looks foreign and rigid. It looks like there is a lot of classes, because in the past we had a few classes but they are typically huge and growing. The number of classes is irrelevant. Usefulness, readability, extensibility, testability, these things are relevant concern.
2) Mapper does a pretty trivial thing: Yes and no. If you watched my presentation you remember I said every layer in Clean has it's own Model object (that includes Entities, which by themselves ARENT just Model objects) and that we should define a rough number of parameters or some guideline when a mapper should be needed (for example, if a class has more than 3 params or it's definitely going to grow, or conversion is complex, lets have a mapper. But the number of times it's going to be used is irrelevant). Mappers are actually extremely useful. Their sole existence is to convert one object into another, or few objects into another and so on and they have very well defined Single Responsibility from SOLID. They make testing of the conversion and it's multiple different variations very easy and also easy to substitute them should we need to. It's also easy to find them by looking for implementation of appropriate protocol. They also leave the users a lot cleaner as the mapping code can be an ugly clutter and also introduce a lot of foreign logic that a class doesn't necessarily need to have